### PR TITLE
doc: Add memory tuning section to user guide

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion-comet/issues/595

### CODE REVIEW

File: docs/source/user-guide/tuning.md [hunk 1]
```diff
@@ -21,6 +21,17 @@ under the License.
 
 Comet provides some tuning options to help you get the best performance from your queries.
 
+## Memory Tuning
+
+Comet currently doesn't share the memory allocation from Spark but owns its own memory allocation.
+That's said, Comet requires additional memory to be allocated. Comet provides some memory related configurations to help you tune the memory usage.
+
+By default, the amount of memory is `spark.comet.memory.overhead.factor` * `spark.executor.memory`.
```
Comment (comphead) [inline docs/source/user-guide/tuning.md:29]: I think we need to specify the precedence as well.
What if both `spark.comet.memory.overhead.factor` and `spark.comet.memoryOverhead` are set ?

File: docs/source/user-guide/tuning.md [hunk 2]
```diff
@@ -21,6 +21,17 @@ under the License.
...
+`COMET_MEMORY_OVERHEAD_FACTOR`'s default value is 0.2. You can increase the factor to require more
```
Comment (andygrove) [inline docs/source/user-guide/tuning.md:30]: Using the internal variable name could cause confusion.

```suggestion
The default value for `spark.comet.memory.overhead.factor` is 0.2. You can increase the factor to require more
```

File: docs/source/user-guide/tuning.md [hunk 3]
```diff
@@ -21,6 +21,17 @@ under the License.
...
+memory for Comet to use, if you see OOM error.
```
Comment (andygrove) [inline docs/source/user-guide/tuning.md:31]: As well as solving OOM, would increasing the memory help with performance?

File: docs/source/user-guide/tuning.md (hunk 1 above)
Comment (viirya) [reply docs/source/user-guide/tuning.md:29]: `spark.comet.memoryOverhead` will be used. Let me add it.

File: docs/source/user-guide/tuning.md (hunk 3 above)
Comment (viirya) [reply docs/source/user-guide/tuning.md]: Added a few sentences about performance impact.

**Review by andygrove — APPROVED** [review]
This is very helpful. Thanks @viirya

**Comment by viirya** [comment]
Merged. Thanks @andygrove @comphead

**Comment by orthoxerox** [comment]
Am I correct in assuming I don't have to increase `spark.executor.memoryOverhead` or `spark.memory.offHeap.size` when using Comet, only `spark.comet.memoryOverhead`? How does this interact with YARN memory management?

**Comment by comphead** [comment]
`spark.comet.memoryOverhead` configures the memory pool size for native execution. it is not the same as `spark.executor.memoryOverhead` or `spark.memory.offHeap.size`.

For YARN memory management, the YARN container respects the `spark.executor.memoryOverhead` and the Comet override this value in `CometDriverPlugin`

```      
sc.conf.set(EXECUTOR_MEMORY_OVERHEAD.key, s"${execMemOverhead + cometMemOverhead}M")
```

**Comment by viirya** [comment]
If Comet `CometPlugin` is configured to be used, it will update Spark executor overhead based on the config value `spark.comet.memoryOverhead` etc. But I don't test it on YARN, not sure Spark plugins comes in before YARN's resource configuration. You probably can test it.

**Comment by orthoxerox** [comment]
@comphead Hm, interesting, I tried launching the container with and without specifying `spark.comet.memoryOverhead` and YARN reported the same amount of memory allocated to my application. I'll check again tomorrow with debug enabled, hopefully I'll see this: https://github.com/apache/datafusion-comet/blob/acb78280e495079312a1067b80e91671852389a5/spark/src/main/scala/org/apache/spark/Plugins.scala#L63C7-L63C14

**Comment by comphead** [comment]
> @comphead Hm, interesting, I tried launching the container with and without specifying `spark.comet.memoryOverhead` and YARN reported the same amount of memory allocated to my application. I'll check again tomorrow with debug enabled, hopefully I'll see this: https://github.com/apache/datafusion-comet/blob/acb78280e495079312a1067b80e91671852389a5/spark/src/main/scala/org/apache/spark/Plugins.scala#L63C7-L63C14

@orthoxerox please try it with applying a separate plugin 
```
--conf spark.plugins=org.apache.spark.CometPlugin
```

**Pull request merged**